pkg/kapis/events/v1alpha2/handler.go

```diff
@@ -0,0 +1,191 @@
+	"strconv"
+	"strings"
+	"time"
+)
+
+type handler struct {
+	tenant tenant.Interface
+	eo     v1alpha2.EventsOperator
+}
+
+func newHandler(factory informers.InformerFactory, ec events.Client) *handler {
+	return &handler{tenant: tenant.New(factory), eo: v1alpha2.NewEventsOperator(ec)}
+}
+
+func (h handler) Query(req *restful.Request, resp *restful.Response) {
+	suser, ok := request.UserFrom(req.Request.Context())
+	if !ok {
+		err := errors.New("cannot obtain user info")
+		klog.Errorln(err)
+		api.HandleForbidden(resp, req, err)
+		return
+	}
+	userWorkspaces, userNamespaces, err := h.userScope(suser)
+	if err != nil {
+		klog.Errorln(err)
+		api.HandleInternalError(resp, req, err)
+		return
+	}
+	filter, err := h.parseFilterFrom(req, userWorkspaces, userNamespaces)
+	if err != nil {
+		klog.Errorln(err)
+		api.HandleInternalError(resp, req, err)
+		return
+	}
+
+	var ar apiEvents.APIResponse
+	if len(filter.InvolvedObjectNamespaceMap) == 0 { // does not query and returns empty result
+		switch req.QueryParameter("operation") {
+		case "histogram":
+			ar.Histogram = &events.Histogram{}
+		case "statistics":
+			ar.Statistics = &events.Statistics{}
+		default:
+			ar.Events = &events.Events{}
+		}
+		resp.WriteAsJson(ar)
+		return
+	}
+
+	switch req.QueryParameter("operation") {
+	case "histogram":
+		interval := req.QueryParameter("interval")
+		if interval == "" {
+			interval = "15m"
+		}
+		ar.Histogram, err = h.eo.CountOverTime(filter, interval)
+	case "statistics":
+		ar.Statistics, err = h.eo.StatisticsOnResources(filter)
+	default:
+		from, _ := strconv.ParseInt(req.QueryParameter("from"), 10, 64)
+		size, err := strconv.ParseInt(req.QueryParameter("size"), 10, 64)
+		if err != nil {
+			size = 10
+		}
+		sorts := map[string]string{
+			"lastTimestamp": req.QueryParameter("sort"),
+		}
+		ar.Events, err = h.eo.SearchEvents(filter, from, size, sorts)
+	}
+
+	if err != nil {
+		klog.Errorln(err)
+		api.HandleInternalError(resp, req, err)
+		return
+	}
+	resp.WriteAsJson(ar)
+}
+
+func (h handler) userScope(user user.Info) (workspaces []*v1alpha1.Workspace, namespaces []*corev1.Namespace, err error) {
+	lw, e := h.tenant.ListWorkspaces(user, query.New())
+	if e != nil {
+		err = fmt.Errorf("error listing workspaces: %s", e)
+		return
+	}
+	for _, r := range lw.Items {
+		if ws, ok := r.(*v1alpha1.Workspace); ok {
+			workspaces = append(workspaces, ws)
+		}
+	}
+	ln, e := h.tenant.ListNamespaces(user, "", query.New())
+	if e != nil {
+		err = fmt.Errorf("error listing namespaces: %s", e)
+		return
+	}
+	for _, r := range ln.Items {
+		if ns, ok := r.(*corev1.Namespace); ok {
+			namespaces = append(namespaces, ns)
+		}
+	}
+	return
+}
+
+func (h handler) parseFilterFrom(req *restful.Request, userWorkspaces []*v1alpha1.Workspace,
+	userNamespaces []*corev1.Namespace) (*events.Filter, error) {
+	queryNamespaces := stringutils.Split(req.QueryParameter("involved_object_namespace_filter"), ",")
```

Comment:
Maybe names like below is more meaningful:

queryNamespaces => namespaceFilter
queryNamespaceSubstrs => namespaceSearching
queryWorkspaces => workspaceFilter
queryWorkspaceSubstrs => workspaceSearching